### ISSUE TITLE
No module named 'fcntl' on Windows 10

### ISSUE THREAD

**Issue by hsluoyz** [wookayin/gpustat#57]
I try to use this tool on Win10. But the used ``fcntl`` library is not available on Win10. Is there any way or plan to support Windows in future?

**Comment by wookayin** [comment]
The `blessings` library has a dependency on `fnctl`; didn't know that. We will try to address this issue soon.

Until then, you can use an older version (< 0.4.0, say v0.3.2) that does not use `fcntl`. It may lack some recent changes or bugfixes, but can you try this?

**Comment by wookayin** [comment]
Following the discussion from the upstream issue erikrose/blessings#21, we can maybe use [this hack](https://stackoverflow.com/questions/1422368/fcntl-substitute-on-windows/25471508#25471508) to work around the issue. As noted, `fcntl.ioctl` is only used to determine the height and weight of the current tty.

**Comment by skjerns** [comment]
is there any update on the dependency on `fcntl`?

**Comment by wookayin** [comment]
We will start using `blessed` (1.17.1+) in place of `blessings`. @skjerns has added a window support in #78, thanks!

Could anyone try `blessed` branch on windows? Are color and `-i` (watch mode) all working well?

**Comment by wookayin** [comment]
@skjerns @hsluoyz @gwzhangx Could you please help verify it now works on windows?

**Comment by skjerns** [comment]
sure:

```
Ran 6 tests in 2.335s

OK
```

**Comment by wookayin** [comment]
@skjerns Thanks. I wonder if the command line program (not only tests) works as expected as test coverage is low (I should improve though) -- like colors, curses-powered features (e.g. `-i` mode). Anyway I'll going to merge it to the master, but let me know if you find any bugs on windows!

**Comment by chanduthedev** [comment]
Hi @wookayin,

I am getting below error on Windows 10 GPU laptop.
`ModuleNotFoundError: No module named 'fcntl'`

I installed by running `pip install gpustat` which installed gpustat v0.6.0 

Any suggestions on how to resolve this issue?
Thanks in advance.

**Comment by wookayin** [comment]
@chanduthedev Have you ever read this issue thread? This was already fixed but yet unreleased. Get the master version: `pip install git+https://github.com/wookayin/gpustat`

**Comment by chanduthedev** [comment]
Hi @wookayin , yes I read. But from the below statement I understand that  gpustat wont support python 2.7 from v1.0. I am using python 2.7, so I installed less than v1.0 version. I am sorry if I interpret incorrectly.

> Note that starting from v1.0, gpustat will support only Python 3.4+. For older versions (python 2.7, <3.4), you can continue using gpustat v0.x.

**Comment by wookayin** [comment]
We do not support python 2.7 anymore. If you want to use python2.7 on windows, please install version 0.3.2. I recommend you to move to python 3.

**Comment by chanduthedev** [comment]
Thanks for quick response. I will try 0.3.2 version once.

**Comment by eusoubrasileiro** [comment]
Just downloaded the master version

Got the following error installing from conda:

`pip install git+https://github.com/wookayin/gpustat`

gpustat --debug
Error on querying NVIDIA devices. Use --debug flag for details
Traceback (most recent call last):
  File "d:\anaconda3\lib\site-packages\pynvml.py", line 641, in _LoadNvmlLibrary
    nvmlLib = CDLL(os.path.join(os.getenv("ProgramFiles", "C:/Program Files"), "NVIDIA Corporation/NVSMI/nvml.dll"))
  File "d:\anaconda3\lib\ctypes\__init__.py", line 364, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] The specified module could not be found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "d:\anaconda3\lib\site-packages\gpustat\cli.py", line 20, in print_gpustat
    gpu_stats = GPUStatCollection.new_query()
  File "d:\anaconda3\lib\site-packages\gpustat\core.py", line 362, in new_query
    N.nvmlInit()
  File "d:\anaconda3\lib\site-packages\pynvml.py", line 608, in nvmlInit
    _LoadNvmlLibrary()
  File "d:\anaconda3\lib\site-packages\pynvml.py", line 646, in _LoadNvmlLibrary
    _nvmlCheckReturn(NVML_ERROR_LIBRARY_NOT_FOUND)
  File "d:\anaconda3\lib\site-packages\pynvml.py", line 310, in _nvmlCheckReturn
    raise NVMLError(ret)
pynvml.NVMLError_LibraryNotFound: NVML Shared Library Not Found

**Comment by wookayin** [comment]
@eusoubrasileiro Your issue is a different one from this. Please feel free to open a new issue.

**Comment by eusoubrasileiro** [comment]
@wookayin  just did that. Hope I can help somehow, deleting here.

**Comment by lintangtimur** [comment]
> @chanduthedev Did you read this issue thread? This was already fixed but yet unreleased. Get the master version: `pip install git+https://github.com/wookayin/gpustat`

when i run `gpustat`
Error on querying NVIDIA devices. Use --debug flag for details

```
Error on querying NVIDIA devices. Use --debug flag for details
Traceback (most recent call last):
  File "c:\python37\lib\site-packages\pynvml\nvml.py", line 731, in _load_nvml_library
    nvml_lib = CDLL(os.path.join(os.getenv("ProgramFiles", "C:/Program Files"), "NVIDIA Corporation/NVSMI/nvml.dll"))
  File "c:\python37\lib\ctypes\__init__.py", line 364, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: [WinError 126] The specified module could not be found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "c:\python37\lib\site-packages\gpustat\cli.py", line 20, in print_gpustat
    gpu_stats = GPUStatCollection.new_query()
  File "c:\python37\lib\site-packages\gpustat\core.py", line 362, in new_query
    N.nvmlInit()
  File "c:\python37\lib\site-packages\pynvml\nvml.py", line 742, in nvmlInit
    _load_nvml_library()
  File "c:\python37\lib\site-packages\pynvml\nvml.py", line 736, in _load_nvml_library
    check_return(NVML_ERROR_LIBRARY_NOT_FOUND)
  File "c:\python37\lib\site-packages\pynvml\nvml.py", line 366, in check_return
    raise NVMLError(ret)
pynvml.nvml.NVMLError_LibraryNotFound: NVML Shared Library Not Found
```

how?

**Comment by wookayin** [comment]
I have made a pre-release (beta) version which contains the fix of this issue. You can grab it as follows:

```
pip install gpustat==1.0.0b1
```

Stable release (1.0.0) will be out in the next few days.

**Comment by h-vetinari** [comment]
> Stable release (1.0.0) will be out in the next few days.

Any update on this?

**Comment by wookayin** [comment]
I just released v1.0.0 to pypi, so from now on `pip install gpustat` will install a stable version which should work well on windows platforms.

**Comment by skjerns** [comment]
whoops, that crashed my computer with a BSOD. It briefly showed the correct output though.
